WcaOnRails/app/webpacker/components/CompetitionOverview/CompetitionFilters.js

```diff
@@ -0,0 +1,662 @@
+  const [searchQuery, setSearchQuery] = useState();
+  const [timeOrder, setTimeOrder] = useState('present');
+  const [displayMode, setDisplayMode] = useState('list');
+
+  const updateCustomCompetitionApiKey = () => {
+    setCompetitionApiKey((prevKey) => ({
+      ...prevKey,
+      sort_by: 'custom',
+      year: `start${customStartDate?.toISOString().split('T')[0] || ''}
+      end${customEndDate?.toISOString().split('T')[0] || ''}`,
+    }));
+  };
+
+  const editSelectedEvents = (eventId) => {
+    setSelectedEvents((prevSelectedEvents) => (
+      prevSelectedEvents.includes(eventId)
+        ? prevSelectedEvents.filter((id) => id !== eventId)
+        : [...prevSelectedEvents, eventId]
+    ));
+  };
+  const editSelectedRegion = (region) => {
+    setSelectedRegion(region === 'all' ? '' : region);
+    setCompetitionApiKey((prevKey) => ({
+      ...prevKey,
+      region: region === 'all' ? '' : region,
+    }));
+  };
+  const editSelectedDelegate = (delegateId) => {
+    setSelectedDelegate(delegateId === 'None' ? '' : delegateId);
+    setCompetitionApiKey((prevKey) => ({
+      ...prevKey,
+      delegate: delegateId === 'None' ? '' : delegateId,
+    }));
+  };
+  const editPastSelectedYear = (newYear) => {
+    setPastSelectedYear(newYear);
+    if (timeOrder === 'past') {
+      setCompetitionApiKey((prevKey) => ({ ...prevKey, sort_by: 'past', year: newYear }));
+    }
+  };
+  const editCustomStartDate = (date) => {
+    setCustomStartDate(date);
+    setCompetitionApiKey((prevKey) => ({
+      ...prevKey,
+      sort_by: 'custom',
+      year: `start${date?.toISOString().split('T')[0] || ''}
+      end${customEndDate?.toISOString().split('T')[0] || ''}`,
+    }));
+  };
+  const editCustomEndDate = (date) => {
+    setCustomEndDate(date);
+    setCompetitionApiKey((prevKey) => ({
+      ...prevKey,
+      sort_by: 'custom',
+      year: `start${customStartDate?.toISOString().split('T')[0] || ''}
+      end${date?.toISOString().split('T')[0] || ''}`,
+    }));
+  };
+  const editSearchQuery = (q) => {
+    setSearchQuery(q);
+    setCompetitionApiKey((prevKey) => ({
+      ...prevKey,
+      search: q,
+    }));
+  };
+  const editTimeOrder = (order) => {
+    if (order === 'past') {
+      setCompetitionApiKey((prevKey) => ({ ...prevKey, sort_by: 'past', year: pastSelectedYear }));
+    } else if (order === 'custom') {
+      // Calling this in a separate function somehow avoids the problem with setState
+      // being asynchronous and causing the query key to be strangely changed back and forth
+      updateCustomCompetitionApiKey();
+    } else {
+      setCompetitionApiKey((prevKey) => ({ ...prevKey, sort_by: order, year: '' }));
+    }
+
+    setTimeOrder(order);
+  };
+
+  const [inProgressComps, setInProgressComps] = useState([]);
+  const [notInProgressFutureComps, setNotInProgressFutureComps] = useState([]);
+  const [recentComps, setRecentComps] = useState([]);
```

Comment:
Are all these states necessary? Can't they be derived from the `competitionsData` object? Less state is usually better.

Reply:
They're probably not necessary. I initially only created `inProgressComps` and `notInProgressFutureComps` just to separate the two and render the two tables for the "present" tab, but it snowballed into having one for every tab. This can most likely be replaced by some more filtering logic in `CompetitionTable.js`.

Reply:
Ah the other thing is I thought switching tabs would look smoother if the comps were stored somewhere and pulled up. After testing how it is with all the `CompetitionTable`s sharing the same state for comps one thing I noticed is that the comps for the previous tab would still be displayed before the comps for the new tab are fetched. This means that the user would not see an empty list with "Loading..." when switching to a new tab, but rather the previous comps lagging behind and then suddenly replaced by the new comps. I wonder if there's some way to circumvent this while keeping the code efficient.

Reply:
> This means that the user would not see an empty list with "Loading..." when switching to a new tab

Ok I think this can be easily solved by temporarily setting the shared state to an empty array whenever `timeOrder` is changed by the reducer.

Reply:
I haven't looked into this much since my last comment, but I think the query has a method to invalidate its data, which could be called whenever the filters update. I'm not sure that's a good solution, but it's something to consider if it works.